[Concurrency] Introduce Actor protocol to which actor classes all conform

DougGregor opened the pull request:
Introduce a new `Actor` protocol, which is a class-bound protocol with only
one requirement:

```
    func enqueue(partialTask: PartialAsyncTask)
```

All actor classes implicitly conform to this protocol, and will synthesize
a definition of `enqueue(partialTask:)` along with storage for the runtime
to manager in the actor.

This design also lets us loosen the constraints on actors somewhat, so that
an actor class can be `@objc` and inherit directly from `NSObject`. Allow this,
and tighten up `@objc` checking to maintain actor isolation across the
language boundary.

DougGregor committed:
[Concurrency] Introduce Actor protocol to which actor classes all conform.

Introduce a new Actor protocol, which is a class-bound protocol with only
one requirement:

    func enqueue(partialTask: PartialAsyncTask)

All actor classes implicitly conform to this protocol, and will synthesize
a (currently empty) definition of `enqueue(partialTask:)` unless a suitable
one is provided explicitly.

theblixguy commented on lib/Sema/DerivedConformanceActor.cpp:
```diff
@@ -0,0 +1,104 @@
+//===--- DerivedConformanceEquatableHashable.cpp - Derived Equatable & co -===//
```
Nit:
```suggestion
//===--- DerivedConformanceActor.cpp - Derived Actor conformance -----------===//
```

DougGregor replied:
Ah, thanks. I made the change locally.

DougGregor committed:
[Concurrency] Fix new source file header.

DougGregor committed:
[Concurrency] Implicitly synthesize actor queue storage and enqueue.

When an actor class has its `enqueue(partialTask:)` implicitly
synthesized, also synthesize a stored property for the actor's queue.
The type of the property is defined by the _Concurrency library
(`_DefaultActorQueue`), and it will be initialized with a call to
`_defaultActorQueueCreate` (also provided by the _Concurrency
library).

Also synthesize the body of the implicitly-generated
`enqueue(partialTask:)`, which will be a call to
`_defaultActorQueueEnqueuePartialTask(actor:queue:partialTask:)`.
Together, all of these allow us to experiment with the form of the
queue and the queue operation without affecting the type checker.

When `enqueue(partialTask:)` is not implicitly synthesized, the queue
storage is not synthesized either. In such cases, the user has taken
over the execution of tasks for the actor, rather than using the
default implementation.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Concurrency] Ensure that enqueue(partialTask:) is first in actor vtables.

Actor classes never have non-actor superclasses, so we can ensure that
all actor classes have a common vtable prefix for the
`enqueue(partialTask:)` operation. This allows us to treat all actor
classes uniformly, without having to go through the Actor witness
table every time.

DougGregor committed:
[Concurrency] Give Actor enqueue(partialTask:) its own ptrauth discriminator.

For actor class's implementations of `enqueue(partialTask:)`, use a
fixed ptrauth discriminator. Paired with the fixed vtable location,
this allows one to invoke this operation on any native actor instance
without knowing the specific type.

DougGregor committed:
[Concurrency] Ban actor-isolated operations from being @objc.

Actor-isolated operations must not be directly accessible from anywhere
that is not already guaranteed to be running within the actor context.
Prevent such operations from being `@objc`, because that would allow
Objective-C code to violate actor isolation.

DougGregor committed:
[Concurrency] Note that 'actor' is a decl modifier, not an attribute.

DougGregor committed:
[Concurrency] Allow actor classes to inherit from NSObject.

NSObject is guaranteed to have no state and no Swift vtable, and is
necessary for Swift classes to implement the NSObject protocol. Allow
it (and only it) as the superclass of an actor class, so that actor
classes can be exposed to Objective-C.

DougGregor commented:
@swift-ci please smoke test

DougGregor committed:
[Concurrency] Actor is not an attribute

DougGregor commented:
@swift-ci please smoke test
